internal/upgrade/upgrade.go

```diff
@@ -93,6 +98,49 @@ func Run(cfg *TestConfig) {
 			).Should(BeTrue())
 		})
 
+		It("successfully finishes control plane nodes rolling update if it is needed", func() {
+			// Check MachinesSpecUpToDate condition on KubeadmControlPlane. Repeat the check 5 times, with some waiting time,
+			// so Cluster API controllers have time to react to upgrade (it is usually instantaneous).
+			numberOfChecks := 5
+			waitBetweenChecks := 5 * time.Second
+			controlPlaneRollingUpdateStarted := false
+
+			for i := 0; i < numberOfChecks; i++ {
+				controlPlane, err := state.GetFramework().GetKubeadmControlPlane(state.GetContext(), cluster.Name, cluster.GetNamespace())
+				Expect(err).NotTo(HaveOccurred())
+
+				if controlPlane == nil {
+					Skip("Control plane resource not found (assuming this is a managed cluster)")
+				}
+
+				if capiconditions.IsFalse(controlPlane, kubeadm.MachinesSpecUpToDateCondition) &&
+					capiconditions.GetReason(controlPlane, kubeadm.MachinesSpecUpToDateCondition) == kubeadm.RollingUpdateInProgressReason {
+					controlPlaneRollingUpdateStarted = true
+					break
+				} else {
+					machinesSpecUpToDateCondition := capiconditions.Get(controlPlane, kubeadm.MachinesSpecUpToDateCondition)
+					if machinesSpecUpToDateCondition == nil {
+						logger.Log("KubeadmControlPlane condition %s is still not set on the KubeadmControlPlane resource, expected condition with Status='False' and Reason='%s'", kubeadm.MachinesSpecUpToDateCondition, kubeadm.RollingUpdateInProgressReason)
+					} else {
+						logger.Log("KubeadmControlPlane condition %s has Status='%s' and Reason='%s', expected condition with Status='False' and Reason='%s'", kubeadm.MachinesSpecUpToDateCondition, machinesSpecUpToDateCondition.Status, machinesSpecUpToDateCondition.Reason, kubeadm.RollingUpdateInProgressReason)
+					}
+				}
+
+				time.Sleep(waitBetweenChecks)
+			}
+
+			if !controlPlaneRollingUpdateStarted {
+				Skip("Control plane nodes rolling update is not happening")
+			}
+
+			mcClient := state.GetFramework().MC()
+			Eventually(
+				wait.IsKubeadmControlPlaneConditionSet(state.GetContext(), mcClient, cluster.Name, cluster.GetNamespace(), kubeadm.MachinesSpecUpToDateCondition, corev1.ConditionTrue, ""),
+				30*time.Minute,
```

Comment:
This is a loooonnnnggggg time to wait in the instance that something went wrong. Is this really a realistic timeframe for this?

Reply:
Probably not really needed this much. In my tests it was around 10-15 minutes. Will put 20 minutes here (15 is a bit of a stretch).

Reply:
Sounds good. We can look at decreasing it in the future if we manage to improve the product to be quicker.

Reply:
Yeah, just not very optimistic here, at least not ATM, as most of that time is on the provider, e.g. getting new machine on AWS, and then on CAPI, e.g. CAPI kubeadm control plane provisioning the nodes.

Reply:
Updated timeout to 20 minutes.